File: ruby/tws.rb
require 'rubygems'
require 'restclient'
require 'open-uri'
require 'active_support/all'

class TWS
  
  attr_accessor :api_key, :api_secret, :stid_host, :stor_host, :stom_host, :stopp_host, :api_version, :default_expire
  
  def initialize opts={}
    self.api_key = opts[:api_key] || ENV['TWS_API_KEY']
    self.api_secret = opts[:api_secret] || ENV['TWS_API_SECRET']
    self.stopp_host = opts[:stopp_host] || "https://stopp.dddws.com"
    self.stom_host = opts[:stom_host] || "https://stom.dddws.com"
    self.stor_host = opts[:stor_host] || "https://stor.dddws.com"
    self.stid_host = opts[:stid_host] || "https://stid.dddws.com"
    self.api_version = opts[:api_version] || 1
    self.default_expire = 1.hour
  end

  def signature string_to_sign
    raise "api_secret is not set" if api_secret.blank?
    CGI.escape(
      Base64.encode64(
        OpenSSL::HMAC.digest(
          OpenSSL::Digest::Digest.new('sha1'),
          api_secret,
          string_to_sign
        )
      ).strip
    )
  end
  
  def expire t = nil
    t = default_expire if t.nil?
    (Time.now + t).to_i
  end
  
  def authenticate
    t = expire
    sig = signature %|POST\n\n#{t}\n/api/v#{api_version}/authenticate|
    response = RestClient.post "#{stid_host}/api/v#{api_version}/authenticate", {
      :key => api_key,
      :signature => sig,
      :method => 'POST',
      :url => "/api/v#{api_version}/authenticate",
      :options => nil,
      :expire => t
    }
    JSON.parse(response.body)
  end
  
  def create_model meta={}, upload_id=nil
    t = expire
    sig = signature %|POST\n\n#{t}\n/api/v#{api_version}/models|
    auth_header = "3WS #{api_key}:#{sig}"
    response = RestClient.post  "#{stor_host}/api/v#{api_version}/models?expire=#{t}",
                                {:meta => meta, :upload_id => upload_id},
                                :Authorization => auth_header
    JSON.parse(response.body)
  end
  
  def get_model id
    t = expire
    sig = signature %|GET\n\n#{t}\n/api/v#{api_version}/models/#{id}|
    auth_header = "3WS #{api_key}:#{sig}"
    response = RestClient.get  "#{stor_host}/api/v#{api_version}/models/#{id}?expire=#{t}",
                                :Authorization => auth_header
    JSON.parse(response.body)
  end
  
  def get_models params={}
    t = expire
    sig = signature %|GET\n\n#{t}\n/api/v#{api_version}/models|
    auth_header = "3WS #{api_key}:#{sig}"
    response = RestClient.get  "#{stor_host}/api/v#{api_version}/models?expire=#{t}&#{params.to_query}",
                                :Authorization => auth_header
    JSON.parse(response.body)
  end
  
  def update_model id, meta={}
    t = expire
    sig = signature %|PUT\n\n#{t}\n/api/v#{api_version}/models/#{id}|
    auth_header = "3WS #{api_key}:#{sig}"
    response = RestClient.put  "#{stor_host}/api/v#{api_version}/models/#{id}?expire=#{t}",
                                {:meta => meta},
                                :Authorization => auth_header
    JSON.parse(response.body)
  end
  
  def delete_model id
    t = expire
    sig = signature %|DELETE\n\n#{t}\n/api/v#{api_version}/models/#{id}|
    auth_header = "3WS #{api_key}:#{sig}"
    response = RestClient.delete  "#{stor_host}/api/v#{api_version}/models/#{id}?expire=#{t}",
                                :Authorization => auth_header
    response.code
  end
  
  def presigned_upload_form starts_with={}, ip=""
    t = expire
    sig = signature %|POST\n\n#{t}\n/api/v#{api_version}/models/presign|
    auth_header = "3WS #{api_key}:#{sig}"
    response = RestClient.post  "#{stor_host}/api/v#{api_version}/models/presign?expire=#{t}",
                                {:ip => ip, :starts_with => starts_with},
                                :Authorization => auth_header
    JSON.parse(response.body)
  end
  
  def upload_model path, opts={}
    presign = presigned_upload_form(opts[:starts_with], opts[:ip])
    begin
      upload_response = RestClient.post presign["form_action"],
                                        presign["form_fields"].merge({ :file => File.new(path), :multipart => true })
    rescue => e
      raise e
    end
    
    raise upload_response.body unless upload_response.code == 204
    
    t = expire
    sig = signature %|POST\n\n#{t}\n/api/v#{api_version}/models|
    auth_header = "3WS #{api_key}:#{sig}"
    response = RestClient.post  "#{stor_host}/api/v#{api_version}/models?expire=#{t}",
                                {:meta => opts[:meta], :upload_id => presign["upload_id"]},
                                :Authorization => auth_header
    JSON.parse(response.body)
  end
  
  def get_link id, filename=nil, expire_time=nil, content_type=nil
    t = expire expire_time
    if filename.nil?
      sig = signature %|GET\n\n#{t}\n/api/v#{api_version}/models/#{id}/download|
      "#{stor_host}/api/v#{api_version}/models/#{id}/download?expire=#{t}&key=#{api_key}&signature=#{sig}&content_type=#{content_type}"
    else
      sig = signature %|GET\n\n#{t}\n/api/v#{api_version}/models/#{id}/#{URI.escape(filename)}|
      "#{stor_host}/api/v#{api_version}/models/#{id}/#{URI.escape(filename)}?expire=#{t}&key=#{api_key}&signature=#{sig}&content_type=#{content_type}"
    end
  end
  
  def request_tptx opts={}
    re = RestClient.post "https://cloud3d.herokuapp.com/api/v1/jobs", {
      :auth_token => 'XyDJkYZ9iwJyeqqaBqWf',
      :input_url => opts[:input_url],
      :input_type => opts[:input_type],
      :include_thumbnail => opts[:include_thumbnail]
    }
    JSON.parse(re.body)
  end
  
  def poll_tptx job_id
    re = RestClient.get "https://cloud3d.herokuapp.com/api/v1/jobs/#{job_id}?auth_token=XyDJkYZ9iwJyeqqaBqWf"
    JSON.parse(re.body)
  end
  
  def get_sessions params={}
    t = expire
    sig = signature %|GET\n\n#{t}\n/api/v#{api_version}/sessions|
    auth_header = "3WS #{api_key}:#{sig}"
    response = RestClient.get  "#{stom_host}/api/v#{api_version}/sessions?expire=#{t}&#{params.to_query}",
                                :Authorization => auth_header
    JSON.parse(response.body)
  end
  
  def get_session id
    t = expire
    sig = signature %|GET\n\n#{t}\n/api/v#{api_version}/sessions/#{id}|
    auth_header = "3WS #{api_key}:#{sig}"
    response = RestClient.get  "#{stom_host}/api/v#{api_version}/sessions/#{id}?expire=#{t}",
                                :Authorization => auth_header
    JSON.parse(response.body)
  end
  
  def create_session timeout=60, engine_version='latest'
    t = expire
    sig = signature %|POST\n\n#{t}\n/api/v#{api_version}/sessions|
    auth_header = "3WS #{api_key}:#{sig}"
    response = RestClient.post  "#{stom_host}/api/v#{api_version}/sessions?expire=#{t}",
                                {:timeout => timeout, :engine_version => engine_version},
                                :Authorization => auth_header
    JSON.parse(response.body)
  end
  
  def close_session id
    t = expire
    sig = signature %|DELETE\n\n#{t}\n/api/v#{api_version}/sessions/#{id}|
    auth_header = "3WS #{api_key}:#{sig}"
    response = RestClient.delete  "#{stom_host}/api/v#{api_version}/sessions/#{id}?expire=#{t}",
                                :Authorization => auth_header
    JSON.parse(response.body)
  end

  def create_run id, platform, code
    t = expire
    sig = signature %|POST\n\n#{t}\n/api/v#{api_version}/sessions/#{id}/runs|
    auth_header = "3WS #{api_key}:#{sig}"
    response = RestClient.post  "#{stom_host}/api/v#{api_version}/sessions/#{id}/runs?expire=#{t}",
                                {:platform => platform, :code => code},
                                :Authorization => auth_header
    JSON.parse(response.body)
  end
  
  def get_runs id, platform=""
    t = expire
    sig = signature %|GET\n\n#{t}\n/api/v#{api_version}/sessions/#{id}/runs|
    auth_header = "3WS #{api_key}:#{sig}"
    response = RestClient.get  "#{stom_host}/api/v#{api_version}/sessions/#{id}/runs?expire=#{t}&platform=#{platform}",
                                :Authorization => auth_header
    JSON.parse(response.body)
  end
  
  def get_run session_id, run_id
    t = expire
    sig = signature %|GET\n\n#{t}\n/api/v#{api_version}/sessions/#{session_id}/runs/#{run_id}|
    auth_header = "3WS #{api_key}:#{sig}"
    response = RestClient.get  "#{stom_host}/api/v#{api_version}/sessions/#{session_id}/runs/#{run_id}?expire=#{t}",
                                :Authorization => auth_header
    JSON.parse(response.body)
  end
  
  # STOPP
  def get_printers params={}
    t = expire
    sig = signature %|GET\n\n#{t}\n/api/v#{api_version}/printers|
    auth_header = "3WS #{api_key}:#{sig}"
    response = RestClient.get "#{stopp_host}/api/v#{api_version}/printers?expire=#{t}&#{params.to_query}", :Authorization => auth_header
    JSON.parse(response.body)
  end
  
  def get_materials params={}
    t = expire
    sig = signature %|GET\n\n#{t}\n/api/v#{api_version}/materials|
    auth_header = "3WS #{api_key}:#{sig}"
    response = RestClient.get "#{stopp_host}/api/v#{api_version}/materials?expire=#{t}&#{params.to_query}", :Authorization => auth_header
    JSON.parse(response.body)
  end
  
  def get_properties params={}
    t = expire
    sig = signature %|GET\n\n#{t}\n/api/v#{api_version}/properties|
    auth_header = "3WS #{api_key}:#{sig}"
    response = RestClient.get "#{stopp_host}/api/v#{api_version}/properties?expire=#{t}&#{params.to_query}", :Authorization => auth_header
    JSON.parse(response.body)
  end
  
end

File: python/tws.py
import cgi
import base64
import hmac
import hashlib
import urllib
import time
import requests
import json
import os

class Tws:
  def __init__( self, api_key = None, api_secret = None,
                stom_host = "https://stom.dddws.com",
                stor_host = "https://stor.dddws.com",
                stid_host = "https://stid.dddws.com",
                api_version = 1 ):

    if not api_key:
      try:
        api_key = os.environ['TWS_API_KEY']
      except:
        pass

    if not api_secret:
      try:
        api_secret = os.environ['TWS_API_SECRET']
      except:
        pass

    self.api_key = api_key
    self.api_secret = api_secret
    self.stom_host = stom_host
    self.stor_host = stor_host
    self.stid_host = stid_host
    self.api_version = api_version
    self.default_expire = 3600
    self.expire = None
  
  def signature(self, string_to_sign):
    try:
      return urllib.parse.quote_plus(base64.b64encode(hmac.new(self.api_secret.encode(),string_to_sign.encode(),hashlib.sha1).digest()).decode())
    except:
      return urllib.quote_plus(base64.b64encode(hmac.new(self.api_secret.encode(),string_to_sign.encode(),hashlib.sha1).digest()).decode())
            
  def set_expire(self, expire_sec = None):
    if expire_sec is None:
      expire_sec = self.default_expire
    self.expire = int(time.time()) + expire_sec
  
  def string_to_sign(self, method, endpoint):
    return "%s\n\n%d\n/api/v%d%s" % (method, self.expire, self.api_version, endpoint)
    
  def authorization(self, method, endpoint):
    return "3WS %s:%s" % (self.api_key, self.signature(self.string_to_sign(method, endpoint)))
  
  def endpoint(self, host, path, extra_params={}):
    try:
      extra = urllib.parse.urlencode(extra_params)
    except:
      extra = urllib.urlencode(extra_params)
    return "%s/api/v%s%s?expire=%s&%s" % (eval("self.%s_host" % host), self.api_version, path, self.expire, extra)
  
  # stid test authentication
  def authenticate(self):
    self.set_expire()
    sig = self.signature(self.string_to_sign('POST', '/authenticate'))
    data = {
      'key': self.api_key,
      'signature': sig,
      'method': 'POST',
      'url': '/api/v1/authenticate',
      'options': '',
      'expire': self.expire
    }
    return requests.post("%s/api/v%d/authenticate" % (self.stid_host, self.api_version), data=data).json()
  
  def create_model(self, meta={}, upload_id=None):
    self.set_expire()
    response = requests.post(
      self.endpoint('stor', '/models'),
      data=json.dumps({"meta": meta, "upload_id": upload_id}),
      headers={'Authorization': self.authorization('POST', '/models'), 'Content-Type': 'application/json'}
    )
    if response.status_code == 201:
      return response.json()
    else:
      return response.status_code

  def get_model(self, mid):
    self.set_expire()
    response = requests.get(
      self.endpoint('stor', '/models/%s' % mid),
      headers={'Authorization': self.authorization('GET', '/models/%s' % mid)}
    )
    if response.status_code == 200:
      return response.json()
    else:
      return response.status_code

  def get_models(self, params={}):
    self.set_expire()
    response = requests.get(
      self.endpoint('stor', '/models', extra_params=params),
      headers={'Authorization': self.authorization('GET', '/models')}
    )
    if response.status_code == 200:
      return response.json()
    else:
      return response.status_code
  
  def update_model(self, mid, meta={}):
    self.set_expire()
    response = requests.put(
      self.endpoint('stor', '/models/%s' % mid),
      data=json.dumps({"meta": meta}),
      headers={'Authorization': self.authorization('PUT', '/models/%s' % mid), 'Content-Type': 'application/json'}
    )
    if response.status_code == 200:
      return response.json()
    else:
      return response.status_code
  
  def delete_model(self, mid):
    self.set_expire()
    response = requests.delete(
      self.endpoint('stor', '/models/%s' % mid),
      headers={'Authorization': self.authorization('DELETE', '/models/%s' % mid)}
    )
    return response.status_code
    
  def presigned_upload_form(self, starts_with={}, ip=""):
    self.set_expire()
    response = requests.post(
      self.endpoint('stor', '/models/presign'),
      data={"starts_with": starts_with, "ip": ip},
      headers = { 'Authorization': self.authorization('POST', '/models/presign') }
    )
    if response.status_code == 200:
      return response.json()
    else:
      return None

  def upload_model(self, path, meta={}, starts_with={}, ip=""):
    self.set_expire()
    presign = self.presigned_upload_form(starts_with, ip)
    files = {'file': open(path, 'rb')}
    upload_response = requests.post(
      presign["form_action"],
      data=presign["form_fields"],
      files=files
    )
    return self.create_model(meta, presign['upload_id'])
    
  def get_link(self, mid, filename=None, expire_sec=None):
    self.set_expire(expire_sec)
    if filename == None:
      sig = self.signature(self.string_to_sign('GET', '/models/%s/download' % mid))
      return "%s/api/v%s/models/%s/download?expire=%s&key=%s&signature=%s" % (self.stor_host, self.api_version, mid, self.expire, self.api_key, sig)
    else:
      try:
        escaped_filename = urllib.parse.quote(filename)
      except:
        escaped_filename = urllib.quote(filename)
      sig = self.signature(self.string_to_sign('GET', '/models/%s/%s' % (mid, escaped_filename)))
      return "%s/api/v%s/models/%s/%s?expire=%s&key=%s&signature=%s" % (self.stor_host, self.api_version, mid, escaped_filename, self.expire, self.api_key, sig)
    
  def get_sessions(self, state=''):
    self.set_expire()
    response = requests.get(
      self.endpoint('stom', '/sessions', extra_params={"state":state}),
      headers = { 'Authorization': self.authorization('GET', '/sessions') }
    )
    if response.status_code == 200:
      return response.json()
    else:
      return response.status_code

  def get_session(self, sid):
    self.set_expire()
    response = requests.get(
      self.endpoint('stom', '/sessions/%s' % sid),
      headers = { 'Authorization': self.authorization('GET', '/sessions/%s' % sid) }
    )
    if response.status_code == 200:
      return response.json()
    else:
      return response.status_code

  def create_session(self, timeout=60, engine_version='20140424'):
    self.set_expire()
    response = requests.post(
      self.endpoint('stom', '/sessions'),
      data = json.dumps({"timeout": timeout, "engine_version": engine_version}),
      headers = { 'Authorization': self.authorization('POST', '/sessions'), 'Content-Type': 'application/json' }
    )
    if response.status_code == 201:
      return response.json()
    else:
      return response.status_code

  def close_session(self, sid):
    self.set_expire()
    response = requests.delete(
      self.endpoint('stom', '/sessions/%s' % sid),
      headers = { 'Authorization': self.authorization('DELETE', '/sessions/%s' % sid) }
    )
    if response.status_code == 200:
      return response.json()
    else:
      return response.status_code

  def create_run(self, sid, platform, code):
    self.set_expire()
    response = requests.post(
      self.endpoint('stom', '/sessions/%s/runs' % sid),
      data = json.dumps({'platform':platform, 'code':code}),
      headers = { 'Authorization': self.authorization('POST', '/sessions/%s/runs' % sid), 'Content-Type': 'application/json' }
    )
    if response.status_code == 201:
      return response.json()
    else:
      return response.status_code

  def get_runs(self, sid, platform=""):
    self.set_expire()
    response = requests.get(
      self.endpoint('stom', '/sessions/%s/runs' % sid, extra_params={'platform': platform}),
      headers = { 'Authorization': self.authorization('GET', '/sessions/%s/runs' % sid)}
    )
    if response.status_code == 200:
      return response.json()
    else:
      return response.status_code
      
  def get_run(self, sid, rid):
    self.set_expire()
    response = requests.get(
      self.endpoint('stom', '/sessions/%s/runs/%s' % (sid, rid)),
      headers = { 'Authorization': self.authorization('GET', '/sessions/%s/runs/%s' % (sid, rid))}
    )
    if response.status_code == 200:
      return response.json()
    else:
      return response.status_code